cadence/contracts/AxelarGateway.cdc

```diff
@@ -0,0 +1,344 @@
+    pre {
+      senderIdentity.check() && senderIdentity.borrow()!.owner != nil: "Cannot borrow reference to SenderIdentity capability"
+    }
+
+    emit ContractCall(
+      sender: senderIdentity.address,
+      destinationChain: destinationChain,
+      destinationContractAddress: destinationContractAddress,
+      payloadHash: String.encodeHex(Crypto.hash(payload, algorithm: HashAlgorithm.KECCAK_256)),
+      payload: payload
+    )
+  }
+
+  access(all) fun isCommandExecuted(commandId: String): Bool {
+    return self.executedCommandIds.contains(commandId)
+  }
+
+  access(all) fun execute(
+    commandIds: [String],
+    commands: [String],
+    params: [[AnyStruct]],
+    operators: [String],
+    weights: [UInt256],
+    threshold: UInt256,
+    signatures: [String]
+  ) {
+    let encodedMessage = self.convertDataToHexEncodedMessage(commandIds: commandIds, commands: commands, params: params)
+
+    var allowOperatorshipTransfer = AxelarAuthWeighted.validateProof(message: encodedMessage, operators: operators, weights: weights, threshold: threshold, signatures: signatures)
+
+    let commandsLength = commandIds.length
+      
+    if (commandsLength != commands.length || commandsLength != params.length) {
+      panic("Invalid Commands")
+    }
+
+    var i = 0
+
+    while i < commandsLength {
+      var commandId = commandIds[i]
+      if (self.executedCommandIds.contains(commandId)) {
+        continue
+      }
+
+      let commandHash = Crypto.hash(commands[i].utf8, algorithm: HashAlgorithm.KECCAK_256)
+      let paramHandler = ParamHandler(params: params[i])
+
+      // Transfer OperatorShip is the only command that is directly processed by this contract - unlike other commannds
+      // it is used for managing gateway operators for the Axelar CGP protocol
+      if (commandHash == self.SELECTOR_TRANSFER_OPERATORSHIP && params[i].length == 3) {
+
+        if (allowOperatorshipTransfer) {
+          var transferOperatorShipParams = paramHandler.generateTransferOperatorshipParams()
+
+          if transferOperatorShipParams != nil {
+
+            // we only need to track this was executed, since there's no other processing for this command
+            self.executedCommandIds.append(commandId)
+            AxelarAuthWeighted.transferOperatorship(message: encodedMessage, operators: operators, weights: weights, threshold: threshold, signatures: signatures, params: transferOperatorShipParams!)
+            emit Executed(commandId: commandId)
+          }
+        }
+      } else if (commandHash == self.SELECTOR_APPROVE_CONTRACT_CALL && params[i].length == 6) {
+        let approveContractCallParams = paramHandler.generateApproveContractCallParams()
+
+        if approveContractCallParams != nil {
+          // store the CGPCommand Resource for later use
+          let oldCgpCommand <- self.approvedCommands[commandId] <- create CGPCommand(commandId: commandId)
```

Comment:
I thought we anticipate commandId values to be unique. Is this just a safety guard?

Reply:
CommandId is unique. However, when I tried to store a new resource into this dictionary using the `insert` method like so `self.approvedCommands.insert(key: commandId, <- create CGPCommand(commandId: commandId))`, I get a `loss of resource` error.

Reply:
I believe that when inserting into a dictionary that store resources, it returns the old resource even if that resource doesn't exist. Therefore, it expects the resource to be processed.

I also tried it this way:
`let oldCgpCommand <- self.approvedCommands.insert(key: commandId, <- create CGPCommand(commandId: commandId))`

but that would be the same implementation as the current one.